src/coreclr/inc/corinfo.h

```diff
@@ -852,7 +852,7 @@ enum CorInfoFlag
     CORINFO_FLG_ARRAY                 = 0x00080000, // class is an array class (initialized differently)
     CORINFO_FLG_OVERLAPPING_FIELDS    = 0x00100000, // struct or class has fields that overlap (aka union)
     CORINFO_FLG_INTERFACE             = 0x00200000, // it is an interface
-    CORINFO_FLG_DONT_DIG_FIELDS       = 0x00400000, // don't ask field info, AOT can't rely on it (used for types outside of AOT compilation version bubble)
+    CORINFO_FLG_DONT_DIG_FIELDS       = 0x00400000, // don't ask field info (used for types outside of AOT compilation version bubble and value arrays)
```

Comment:
This flag was called `CORINFO_FLG_DONT_PROMOTE` at the time of the prototype. We do not want inline arrays to promote as indexing relies on contiguous layout.

I assume the flag still has the effect of suppressing field promotion.

Reply:
We should double check with the JIT team - I would think we should be setting similar flags that we set for explicit layout structs and explicit layout structs don't set this one.

Reply:
@dotnet/jit-contrib can someone from the JIT team take a look at this part?

Reply:
`CORINFO_FLG_DONT_DIG_FIELDS` doesn't always suppress promotion (eg a single gc ref field struct will still get promoted).

I would introduce a new flag describing this, something like `CORINFO_FLG_INDEXABLE_FIELDS`, and update the jit to block promotion when it sees it.  Looks like there is still one free bit?

Down the road we might be doing more flexible partial promotion and perhaps even handle cases like these where we know exactly which fields are being touched over some span of code.

Reply:
> update the jit to block promotion when it sees it

Is there an easy way to find all the places that need to care about the bit?

Reply:
`StructPromotionHelper::CanPromoteStructType` is the only place, promotion-wise.

Though it is not clear to me why we need to block the promotion that `CORINFO_FLG_DONT_DIG_FIELDS` allows; it should be ok to promote `object[1]` as `object`.

Reply:
> ... it should be ok to promote...

Right, down the road we can allow promotion in some cases (single field, all accesses with known indices, etc), but simplest for now to disallow.

Reply:
Does this need a new flag? I would expect the same as @MichalStrehovsky said:

> I would think we should be setting similar flags that we set for explicit layout structs and explicit layout structs don't set this one.

I.e. specifying the same custom layout flags would have the desired effect (and maybe even be less conservative than just disabling promotion -- I think we might be able to dependently promote these then?).

Reply:
These structs only have one field so we'd somehow need to know that that field has special meaning if we're going to do things like allow dependent promotion. If all we know is that there is explicit layout; we'd have to infer this from there just being one field at offset zero with unusual size.

I'd rather know exactly what kind of struct we are looking at.

Reply:
We're not going to be able to track any field but the first, but we can dependently promote the entire thing and track the first field assuming none of it is address exposed. It is what happens for a struct like:
```csharp
[StructLayout(LayoutKind.Explicit, Size = 32)]
private struct Foo
{
    [FieldOffset(0)]
    public int A;
}
```

Reply:
I think it makes more sense to be conservative at the beginning as we have very little real-world coverage of the feature and there will be some learning of what corner cases may result from its use. 

Using a distinct flag seems reasonable. There are still a few unused flags left, so there is no shortage.

Reply:
I agree being conservative is fine, but that's just a side benefit. What I think is unfortunate is bloating the JIT-EE interface with more things and bumping the JIT-EE GUID if we can avoid it.

If you decide to leave this as-is then please make sure you bump the JIT-EE GUID, maybe I'm missing it but I don't see that in this PR.

For testing -- how hard would it be to begin consuming the support? Once I can consume it through Roslyn it should not be hard to add support in Fuzzlyn that will get us at least some coverage of different patterns of usage.

Reply:
@VSadov I do think we need a JIT-EE GUID update here -- otherwise it can expose all sorts of latent issues for the JIT team once we get newer collections with this flag that older JITs do not handle.

Reply:
Makes sense. Thanks!! I will change the guid in a follow up change.

In theory the presence of the attribute will imply the new runtime and prevent many versioning issues, but it is better to be sure.

Reply:
> how hard would it be to begin consuming the support? 

Fairly easy. Roslyn will add a layer of convenience, but otherwise the feature can be used directly. 
I’ve added some use cases in corlib and libraries as a part of the change.

The only usual caveat with using a feature this early is possibly having to change things if the feature changes based on feedback.

Reply:
@jakobbotsch   Sorry, did not notice your yesterdays comment.

Reply:
> In theory the presence of the attribute will imply the new runtime and prevent many versioning issues, but it is better to be sure.

The problem for the JIT team is that we disambiguate compatibility of SPMI collections via the JIT-EE GUID. We automatically collect every Sunday, so after that we will start seeing contexts with the new flag set and no indication that newer JITs are needed to properly handle this. That might not cause any problems except silent bad codegen for our SPMI runs, but it could also conceivably cause spurious assertion failures.

FWIW, @EgorBo just merged #83430 which includes a JIT-EE GUID bump, so you won't need to do that in a follow-up change after all.

> Fairly easy. Roslyn will add a layer of convenience, but otherwise the feature can be used directly.
> I’ve added some use cases in corlib and libraries as a part of the change.

I've looked at some of the examples, that looks fairly straightforward. I guess the main thing is how to get access to `InlineArrayAttribute` from the programs I'm generating.